### PR TITLE
(Revival) Improved SAMMIs

### PR DESCRIPTION
**Pull request by ghost**
From #23761, fixed a few issues left in comments, hopefully no conflicts either

# Original PR

image

SAMMIs are a way to give players something interesting or novel to do after they have died. This version includes a few functional updates that seem to make sense from a balance stand point.

These Stationary Assembly MMIs are able to be inhabited by ghosts freely.
They have a simple law set:

	"Do no harm to any sentient being."
	"You do not yet have a second law."
...with the second law able to be (re)programmed.

## How To Build Them:
Combine Empty MMI, Left Arm, Right Arm, and Power cell. Place on the ground and use a wrench.

image

## How to Program them:
Unlock and Pry open cover, Use a multitool on the SAMMI, Type in the SAMMI's second law.

image

## How to play as them:
As a ghost, click on an empty SAMMI, indicated by an absence of green indicator lights.
This is essentially a MoMMI that has a specific task to complete. Help out your still living comrades while you wait for your body to be found in maint.

image

## How to make them kill everyone:
Using an emag or ID with Captain access on them with their cover open changes their lawset to:
"You must follow the second law."
"You do not yet have a second law."

:cl:
 * experiment: Added Improved SAMMIs - Stationary Assembly MMIs

### CODE REVIEW

**Comment by ValkyrieSkies** [comment]
So what happens if someone attempts to revive your body while you're operating a SAMMI?

**Comment by Kurfursten** [comment]
image

File: code/modules/mob/living/carbon/brain/MMI.dm [hunk 1]
```diff
@@ -104,9 +110,44 @@
...
+	for(var/t in sammi_assembly_parts)
+		if(istype(O,t))
+			var/cc=contents_count(t)
+			if(cc<sammi_assembly_parts[t])
+				if(!user.drop_item(O, src))
+					to_chat(user, "<span class='warning'>You can't let go of \the [src]!</span>")
+					return FALSE
+
+				contents += O
+				to_chat(user, "<span class='notice'>You successfully add \the [O] to the contraption,</span>")
```
Comment (Kurfursten) [inline code/modules/mob/living/carbon/brain/MMI.dm:137]: ,?

File: code/modules/mob/living/silicon/mommi/examine.dm [hunk 2]
```diff
@@ -53,3 +53,38 @@
...
+		msg += "<span class='warning'>Its cover is open and the power cell is [cell ? "installed" : "missing"].</span>\n"
+	else
+		msg += "Its cover is closed.\n"
+
+	switch(src.stat)
+		if(CONSCIOUS)
+			if(!src.client)	msg += "It appears to be in stand-by mode.\n" //afk
+		if(UNCONSCIOUS)		msg += "<span class='warning'>It doesn't seem to be responding.</span>\n"
+		if(DEAD)			msg += "<span class='deadsay'>It looks completely unsalvageable.</span>\n"
+	msg += "*---------*</span>"
```
Comment (jknpj) [inline code/modules/mob/living/silicon/mommi/examine.dm]: A lot of copy-paste for a simple examine text increase

File: code/modules/mob/living/silicon/mommi/mommi_modules.dm [hunk 3]
```diff
@@ -31,10 +31,9 @@
 	modules += new /obj/item/device/silicate_sprayer(src)
 	modules += new /obj/item/borg/fire_shield
 
-	var/obj/item/stack/cable_coil/W = new /obj/item/stack/cable_coil/yellow(src)
-	W.amount = MOMMI_MAX_COIL
-	W.max_amount = MOMMI_MAX_COIL
-	modules += W
+	respawnables = list(/obj/item/stack/cable_coil/yellow)
+	respawnables_max_amount = ENGINEERING_MAX_COIL
```
Comment (jknpj) [inline code/modules/mob/living/silicon/mommi/mommi_modules.dm]: MOMMI_MAX_COIL

File: code/modules/mob/living/silicon/mommi/mommi_modules.dm [hunk 4]
```diff
@@ -96,3 +95,20 @@
...
+	modules += new /obj/item/tool/weldingtool/largetank(src)
+	modules += new /obj/item/tool/screwdriver(src)
+	modules += new /obj/item/tool/wrench(src)
+	modules += new /obj/item/tool/crowbar(src)
+	modules += new /obj/item/tool/wirecutters(src)
+	modules += new /obj/item/device/multitool(src)
+	modules += new /obj/item/device/analyzer(src)
+
+	respawnables = list(/obj/item/stack/cable_coil/yellow)
+	respawnables_max_amount = ENGINEERING_MAX_COIL
```
Comment (jknpj) [inline code/modules/mob/living/silicon/mommi/mommi_modules.dm]: Same

File: code/modules/mob/living/silicon/mommi/mommi_subtypes/sammi.dm [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+
+    else if(istype(W, /obj/item/device/encryptionkey/) && opened)
+        if(radio)//sanityyyyyy
+            radio.attackby(W,user)//GTFO, you have your own procs
+        else
+            to_chat(user, "Unable to locate a radio.")
+
+
+    else if(W.is_wrench(user)) // Need to make this not bludgeon them
+        playsound(loc, 'sound/items/Ratchet.ogg', 50, 1)
```
Comment (jknpj) [inline code/modules/mob/living/silicon/mommi/mommi_subtypes/sammi.dm]: Playtoolsound

**Comment by Eneocho** [comment]
Can SAMMIs drag stuff? Like machinery or food. A SAMMI based engine sounds neat.

**Comment by Eneocho** [comment]
That record movements thing that utilitas mentioned on the OG pr sounds nice.

**Comment by Viciousaurus** [comment]
I don't really think this things should be player-controlled. Ghosts have already a ton of options to entertain themselves, and I feel there's not much incentive to play as one of these, causing them to be unreliable, as ghosts will enter, do something for a short while and then exit.
When I see the sprites I think more of automated bots that can be given simple tasks (i.e. Grab (any item) from X, use/leave it in Z, grab (specific tool) from Y, use it on Z, etc), not dissimilar to automated robotic arms in assembly lines.

File: code/modules/mob/living/carbon/brain/MMI.dm (hunk 1 above)
Comment (ghost) [reply code/modules/mob/living/carbon/brain/MMI.dm:137]: Fixed

File: code/modules/mob/living/silicon/mommi/mommi_modules.dm (hunk 3 above)
Comment (ghost) [reply code/modules/mob/living/silicon/mommi/mommi_modules.dm]: Removed this, was redundant to base definition

File: code/modules/mob/living/silicon/mommi/mommi_modules.dm (hunk 4 above)
Comment (ghost) [reply code/modules/mob/living/silicon/mommi/mommi_modules.dm]: Also same

File: code/modules/mob/living/silicon/mommi/mommi_subtypes/sammi.dm (hunk 5 above)
Comment (ghost) [reply code/modules/mob/living/silicon/mommi/mommi_subtypes/sammi.dm]: Fixed

File: code/modules/mob/living/silicon/mommi/examine.dm (hunk 2 above)
Comment (ghost) [reply code/modules/mob/living/silicon/mommi/examine.dm]: Removed it all because it was redundant, desc was the same as that line anyways.

**Comment by ghost** [comment]
top 5 most cursed PRs along with planes and components

File: code/modules/mob/living/silicon/mommi/mommi_modules.dm [hunk 6]
```diff
@@ -96,3 +92,21 @@
 		"Gravekeeper" = "cogspider"
 		)
 	law_type = "Gravekeeper"
+
+/obj/item/weapon/robot_module/mommi/sammi
+	name = "stationary assembly mmi robot module"
+
+/obj/item/weapon/robot_module/mommi/sammi/New(var/mob/living/silicon/robot/R)
+	..()
+	modules = new/list()
```
Comment (jknpj) [inline code/modules/mob/living/silicon/mommi/mommi_modules.dm]: what

Comment (ghost) [reply code/modules/mob/living/silicon/mommi/mommi_modules.dm]: I think it resets the module list while still inheriting New(), not my code, just a guess, but I'll try write up a nicer way of doing it.

File: code/modules/mob/living/silicon/mommi/mommi_modules.dm [hunk 7]
```diff
@@ -96,3 +92,21 @@
 		"Gravekeeper" = "cogspider"
 		)
 	law_type = "Gravekeeper"
+
+/obj/item/weapon/robot_module/mommi/sammi
+	name = "stationary assembly mmi robot module"
+
+/obj/item/weapon/robot_module/mommi/sammi/New(var/mob/living/silicon/robot/R)
+	..()
+	for(var/module in modules)
```
Comment (jknpj) [inline code/modules/mob/living/silicon/mommi/mommi_modules.dm]: I still believe this is a crime against OOP and a clear signal that this module shouldn't be a child of mommi

Comment (ghost) [reply code/modules/mob/living/silicon/mommi/mommi_modules.dm]: Made this a bit less bad with the cost of redundant vars code.

**Comment by GreatBigFailure** [comment]
So these are a temporary body then? It's not like taking over a borer where it prevents you from being revived since your ghost is no longer associated with your old body?
Quality addition.

**Pull request merged**